app/clients/sms/mmg.py

```diff
@@ -77,22 +77,6 @@ def init_app(self, current_app, statsd_client, *args, **kwargs):
         self.statsd_client = statsd_client
         self.mmg_url = current_app.config.get('MMG_URL')
 
-    def record_outcome(self, success, response):
-        status_code = response.status_code if response else 503
```

Comment:
we're losing the logging of this information - what the status code is is quite useful. Other things such as the response body might also be useful but we don't currently capture that at all.

(You can see by looking at existing log lines that we don't currently log this exception further up. eg this kibana link[^1] for a firetext 503 the other week


[^1]: https://kibana.logit.io/s/9423a789-282c-4113-908d-0be3b1bc9d1d/app/kibana#/discover?_g=(refreshInterval:(display:Off,pause:!f,value:0),time:(from:now-7d,mode:quick,to:now))&_a=(columns:!(message),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'logstash-*',key:requestId,negate:!f,type:phrase,value:'216c191853a5f7c77da9d9e2ee056025'),query:(match:(requestId:(query:'216c191853a5f7c77da9d9e2ee056025',type:phrase)))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'logstash-*',key:cf.app,negate:!f,type:phrase,value:notify-delivery-worker-sender),query:(match:(cf.app:(query:notify-delivery-worker-sender,type:phrase))))),index:'logstash-*',interval:auto,query:(query_string:(analyze_wildcard:!t,query:'*')),sort:!('@timestamp',desc))

Reply:
Yeah, I noticed. I was hoping [this commit](https://github.com/alphagov/notifications-api/pull/3493/commits/2746be97c4f5b8266c03f7aeb3ac1c1d76e0559d) would be a reasonable / better replacement over our custom interpretation of the status code. Having the response at this level is tricky because it couples the base class to the way we make the request down below. We can of course move / have some logs back down - that commit is my attempt to have the best of both. What do you think?

Reply:
Ah, good point. I'd missed that the `ReachClientResponseException` overrides the base exc class's `__str__` to contain the code. (And so does MMGClientResponseException and FiretextClientResponseException - those three exception class implementations look like a good candidate for DRYing out 🤔)

I like the approach you've added in provider_tasks. i think logging exceptions at the point that they're caught/suppressed is generally best practice.

i just tested it and the log line looks like the following on a 504, which i think is fine.

```
[2022-03-29 11:22:45,703: WARNING/ForkPoolWorker-2] SMS notification delivery for id: a2ff3cc8-2bd8-4ffd-ab1f-4886843417ab failed
Traceback (most recent call last):
  File "/home/vcap/app/app/clients/sms/reach.py", line 40, in try_send_sms
    response = request(
  File "/usr/local/lib/python3.9/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 528, in request
    prep = self.prepare_request(req)
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 456, in prepare_request
    p.prepare(
  File "/usr/local/lib/python3.9/site-packages/requests/models.py", line 316, in prepare
    self.prepare_url(url, params)
  File "/usr/local/lib/python3.9/site-packages/requests/models.py", line 390, in prepare_url
    raise MissingSchema(error)
requests.exceptions.MissingSchema: Invalid URL 'TODO': No schema supplied. Perhaps you meant http://TODO?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/vcap/app/app/celery/provider_tasks.py", line 23, in deliver_sms
    send_to_providers.send_sms_to_provider(notification)
  File "/home/vcap/app/app/delivery/send_to_providers.py", line 86, in send_sms_to_provider
    raise e
  File "/home/vcap/app/app/delivery/send_to_providers.py", line 81, in send_sms_to_provider
    provider.send_sms(**send_sms_kwargs)
  File "/home/vcap/app/app/clients/sms/__init__.py", line 50, in send_sms
    raise e
  File "/home/vcap/app/app/clients/sms/__init__.py", line 46, in send_sms
    response = self.try_send_sms(to, content, reference, international, sender)
  File "/home/vcap/app/app/clients/sms/reach.py", line 56, in try_send_sms
    raise ReachClientResponseException(response=e.response, exception=e)
app.clients.sms.reach.ReachClientResponseException: Code 504 text Gateway Time-out exception Invalid URL 'TODO': No schema supplied. Perhaps you meant http://TODO?
```

so in summary: i think your approach is fine and losing that line from record_outcome is fine now that we print out the stack trace too.

as an aside it might be nice to DRY up the three SMS exception classes since the functions look identical to me and all override the base class

Reply:
Thank you 🙂.

I had the same thought about the exception classes (and the JSON validation). In theory - and in practice for Firetext - they _could_ be different so the duplication seemed justified for the validation at least.

For the exceptions, how do you feel about this?

https://github.com/alphagov/notifications-api/pull/3493/commits/0fa34339353e2e20b8fa2adf2e252307cf48d763

Reply:
ooh not bad, didn't think about removing the subclasses entirely. i was initially a little worried that it might not contain enough cues as to which provider had the issues, but the stack trace includes the provider file:

```
[...]
  File "/home/vcap/app/app/clients/sms/reach.py", line 50, in try_send_sms
    raise SmsClientResponseException("Request failed")
```

(plus also there's the separate log line `Provider request for reach failed` but that doesn't have any context)